File: 计算机图形学/opencv/15_opencv_center_rotation.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <cmath>

using namespace cv;
using namespace std;

int main()
{
    /*
     * 使用OpenCV打开摄像头并实时展示画面，同时进行蓝色积木块的颜色分拣
     * 增强功能：显示真实边框、中心点坐标和旋转角度
     */
    
    // 创建VideoCapture对象，参数2表示摄像头ID
    VideoCapture cap(2); // 如果电脑上有多个摄像头，需要调整摄像头的id，默认值0
    
    // 检查摄像头是否成功打开
    if (!cap.isOpened()) {
        cout << "错误：无法打开摄像头" << endl;
        return -1;
    }
    
    cout << "摄像头已打开，蓝色积木块检测程序启动（增强版）" << endl;
    cout << "功能：真实边框包裹 + 中心点坐标 + 旋转角度" << endl;
    cout << "按 'q' 键退出" << endl;
    
    // 定义蓝色在HSV颜色空间中的范围
    Scalar lower_blue(100, 50, 50);   // 蓝色下限
    Scalar upper_blue(130, 255, 255); // 蓝色上限
    
    while (true) {
        Mat frame;
        // 读取一帧画面
        bool ret = cap.read(frame);
        
        // 检查是否成功读取到画面
        if (!ret) {
            cout << "错误：无法读取摄像头画面" << endl;
            break;
        }
        
        // 获取原始画面的尺寸
        int height = frame.rows;
        int width = frame.cols;
        
        // 将画面缩小为原来的1/2
        int new_width = width / 2;
        int new_height = height / 2;
        Mat resized_frame;
        resize(frame, resized_frame, Size(new_width, new_height));
        
        // 将BGR颜色空间转换为HSV颜色空间
        Mat hsv;
        cvtColor(resized_frame, hsv, COLOR_BGR2HSV);
        
        // 创建蓝色掩码
        Mat mask;
        inRange(hsv, lower_blue, upper_blue, mask);
        
        // 进行形态学操作，去除噪声
        Mat kernel = getStructuringElement(MORPH_RECT, Size(5, 5));
        morphologyEx(mask, mask, MORPH_OPEN, kernel);
        morphologyEx(mask, mask, MORPH_CLOSE, kernel);
        
        // 寻找轮廓
        vector<vector<Point>> contours;
        vector<Vec4i> hierarchy;
        findContours(mask, contours, hierarchy, RETR_EXTERNAL, CHAIN_APPROX_SIMPLE);
        
        // 创建结果图像的副本
        Mat result_frame = resized_frame.clone();
        
        // 处理检测到的轮廓
        int blue_objects_count = 0;
        for (size_t i = 0; i < contours.size(); i++) {
            // 计算轮廓面积，过滤掉太小的区域
            double area = contourArea(contours[i]);
            if (area > 500) { // 最小面积阈值
                blue_objects_count++;
                
                // 方法1：绘制轮廓的真实边框（轮廓本身）
                drawContours(result_frame, contours, (int)i, Scalar(0, 255, 0), 2);
                
                // 方法2：计算最小外接矩形（带旋转角度）
                RotatedRect rect = minAreaRect(contours[i]);
                Point2f vertices[4];
                rect.points(vertices);
                
                // 将Point2f转换为Point并绘制最小外接矩形
                vector<Point> box_points;
                for (int j = 0; j < 4; j++) {
                    box_points.push_back(Point((int)vertices[j].x, (int)vertices[j].y));
                }
                vector<vector<Point>> box_contour = {box_points};
                drawContours(result_frame, box_contour, 0, Scalar(255, 0, 0), 2);
                
                // 获取矩形的中心点、尺寸和角度
                Point2f center = rect.center;
                Size2f size = rect.size;
                float angle = rect.angle;
                int center_x = (int)center.x;
                int center_y = (int)center.y;
                
                // 计算轮廓的质心（更精确的中心点）
                Moments M = moments(contours[i]);
                int cx, cy;
                if (M.m00 != 0) {
                    cx = (int)(M.m10 / M.m00);
                    cy = (int)(M.m01 / M.m00);
                } else {
                    cx = center_x;
                    cy = center_y;
                }
                
                // 绘制中心点
                circle(result_frame, Point(cx, cy), 8, Scalar(0, 0, 255), -1);
                circle(result_frame, Point(cx, cy), 12, Scalar(255, 255, 255), 2);
                
                // 绘制角度指示线
                // 将角度转换为弧度
                double angle_rad = angle * M_PI / 180.0;
                int line_length = 50;
                
                // 计算指示线的终点
                int end_x = (int)(cx + line_length * cos(angle_rad));
                int end_y = (int)(cy + line_length * sin(angle_rad));
                
                // 绘制角度指示线
                arrowedLine(result_frame, Point(cx, cy), Point(end_x, end_y), Scalar(255, 255, 0), 3);
                
                // 添加文本标签
                int label_y_offset = 0;
                
                // 显示积木块编号
                string label = "Block " + to_string(blue_objects_count);
                putText(result_frame, label, Point(cx - 50, cy - 60 + label_y_offset), 
                       FONT_HERSHEY_SIMPLEX, 0.6, Scalar(255, 255, 255), 2);
                label_y_offset += 20;
                
                // 显示中心点坐标
                string center_text = "Center: (" + to_string(cx) + ", " + to_string(cy) + ")";
                putText(result_frame, center_text, Point(cx - 50, cy - 60 + label_y_offset), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                label_y_offset += 15;
                
                // 显示旋转角度
                string angle_text = "Angle: " + to_string((int)angle) + "°";
                putText(result_frame, angle_text, Point(cx - 50, cy - 60 + label_y_offset), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                label_y_offset += 15;
                
                // 显示面积
                string area_text = "Area: " + to_string((int)area);
                putText(result_frame, area_text, Point(cx - 50, cy - 60 + label_y_offset), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                
                // 在图像左上角显示详细信息
                int info_start_y = 30 + (blue_objects_count - 1) * 80;
                string block_header = "=== Block " + to_string(blue_objects_count) + " ===";
                putText(result_frame, block_header, Point(10, info_start_y), 
                       FONT_HERSHEY_SIMPLEX, 0.6, Scalar(0, 255, 255), 2);
                
                string center_info = "Center: (" + to_string(cx) + ", " + to_string(cy) + ")";
                putText(result_frame, center_info, Point(10, info_start_y + 20), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                
                string rotation_info = "Rotation: " + to_string((int)angle) + "°";
                putText(result_frame, rotation_info, Point(10, info_start_y + 35), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                
                string area_info = "Area: " + to_string((int)area) + " px²";
                putText(result_frame, area_info, Point(10, info_start_y + 50), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
                
                string size_info = "Size: " + to_string((int)size.width) + "x" + to_string((int)size.height);
                putText(result_frame, size_info, Point(10, info_start_y + 65), 
                       FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 255, 255), 1);
            }
        }
        
        // 在图像底部显示检测到的蓝色积木块总数
        string total_info = "Total Blue Blocks Detected: " + to_string(blue_objects_count);
        putText(result_frame, total_info, Point(10, result_frame.rows - 20), 
               FONT_HERSHEY_SIMPLEX, 0.7, Scalar(0, 255, 0), 2);
        
        // 添加图例说明
        int legend_x = result_frame.cols - 250;
        putText(result_frame, "Legend:", Point(legend_x, 30), 
               FONT_HERSHEY_SIMPLEX, 0.6, Scalar(255, 255, 255), 2);
        putText(result_frame, "Green: Contour", Point(legend_x, 50), 
               FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 255, 0), 1);
        putText(result_frame, "Blue: Min Area Rect", Point(legend_x, 65), 
               FONT_HERSHEY_SIMPLEX, 0.5, Scalar(255, 0, 0), 1);
        putText(result_frame, "Red: Center Point", Point(legend_x, 80), 
               FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 0, 255), 1);
        putText(result_frame, "Yellow: Rotation", Point(legend_x, 95), 
               FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0, 255, 255), 1);
        
        // 显示原始画面和处理结果
        imshow("Blue Block Detection - Enhanced", result_frame);
        imshow("Blue Mask", mask);
        
        // 检查按键输入，如果按下 'q' 键则退出
        char key = waitKey(1) & 0xFF;
        if (key == 'q') {
            cout << "用户按下 'q' 键，退出程序" << endl;
            break;
        }
    }
    
    // 释放摄像头资源
    cap.release();
    // 关闭所有OpenCV窗口
    destroyAllWindows();
    cout << "摄像头已关闭，蓝色积木块检测程序结束" << endl;
    
    return 0;
}

File: 计算机图形学/opencv/15_opencv_center_rotation.py
import cv2
import numpy as np
import math

def main():
    """
    使用OpenCV打开摄像头并实时展示画面，同时进行蓝色积木块的颜色分拣
    增强功能：显示真实边框、中心点坐标和旋转角度
    """
    # 创建VideoCapture对象，参数0表示默认摄像头
    cap = cv2.VideoCapture(2) #如果电脑上有多个摄像头，需要调整摄像头的id，默认值0
    
    # 检查摄像头是否成功打开
    if not cap.isOpened():
        print("错误：无法打开摄像头")
        return
    
    print("摄像头已打开，蓝色积木块检测程序启动（增强版）")
    print("功能：真实边框包裹 + 中心点坐标 + 旋转角度")
    print("按 'q' 键退出")
    
    # 定义蓝色在HSV颜色空间中的范围
    lower_blue = np.array([100, 50, 50])   # 蓝色下限
    upper_blue = np.array([130, 255, 255]) # 蓝色上限
    
    while True:
        # 读取一帧画面
        ret, frame = cap.read()
        
        # 检查是否成功读取到画面
        if not ret:
            print("错误：无法读取摄像头画面")
            break
        
        # 获取原始画面的尺寸
        height, width = frame.shape[:2]
        
        # 将画面缩小为原来的1/2
        new_width = width // 2
        new_height = height // 2
        resized_frame = cv2.resize(frame, (new_width, new_height))
        
        # 将BGR颜色空间转换为HSV颜色空间
        hsv = cv2.cvtColor(resized_frame, cv2.COLOR_BGR2HSV)
        
        # 创建蓝色掩码
        mask = cv2.inRange(hsv, lower_blue, upper_blue)
        
        # 进行形态学操作，去除噪声
        kernel = np.ones((5,5), np.uint8)
        mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
        mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)
        
        # 寻找轮廓
        contours, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
        
        # 创建结果图像的副本
        result_frame = resized_frame.copy()
        
        # 处理检测到的轮廓
        blue_objects_count = 0
        for contour in contours:
            # 计算轮廓面积，过滤掉太小的区域
            area = cv2.contourArea(contour)
            if area > 500:  # 最小面积阈值
                blue_objects_count += 1
                
                # 方法1：绘制轮廓的真实边框（轮廓本身）
                cv2.drawContours(result_frame, [contour], -1, (0, 255, 0), 2)
                
                # 方法2：计算最小外接矩形（带旋转角度）
                rect = cv2.minAreaRect(contour)
                box = cv2.boxPoints(rect)
                box = np.int32(box)
                
                # 绘制最小外接矩形
                cv2.drawContours(result_frame, [box], 0, (255, 0, 0), 2)
                
                # 获取矩形的中心点、尺寸和角度
                center, (width_rect, height_rect), angle = rect
                center_x, center_y = int(center[0]), int(center[1])
                
                # 计算轮廓的质心（更精确的中心点）
                M = cv2.moments(contour)
                if M["m00"] != 0:
                    cx = int(M["m10"] / M["m00"])
                    cy = int(M["m01"] / M["m00"])
                else:
                    cx, cy = center_x, center_y
                
                # 绘制中心点
                cv2.circle(result_frame, (cx, cy), 8, (0, 0, 255), -1)
                cv2.circle(result_frame, (cx, cy), 12, (255, 255, 255), 2)
                
                # 绘制角度指示线
                # 将角度转换为弧度
                angle_rad = math.radians(angle)
                line_length = 50
                
                # 计算指示线的终点
                end_x = int(cx + line_length * math.cos(angle_rad))
                end_y = int(cy + line_length * math.sin(angle_rad))
                
                # 绘制角度指示线
                cv2.arrowedLine(result_frame, (cx, cy), (end_x, end_y), (255, 255, 0), 3)
                
                # 添加文本标签
                label_y_offset = 0
                
                # 显示积木块编号
                label = f"Block {blue_objects_count}"
                cv2.putText(result_frame, label, (cx - 50, cy - 60 + label_y_offset), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)
                label_y_offset += 20
                
                # 显示中心点坐标
                center_text = f"Center: ({cx}, {cy})"
                cv2.putText(result_frame, center_text, (cx - 50, cy - 60 + label_y_offset), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                label_y_offset += 15
                
                # 显示旋转角度
                angle_text = f"Angle: {angle:.1f}°"
                cv2.putText(result_frame, angle_text, (cx - 50, cy - 60 + label_y_offset), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                label_y_offset += 15
                
                # 显示面积
                area_text = f"Area: {int(area)}"
                cv2.putText(result_frame, area_text, (cx - 50, cy - 60 + label_y_offset), 
                           cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                
                # 在图像左上角显示详细信息
                info_start_y = 30 + (blue_objects_count - 1) * 80
                cv2.putText(result_frame, f"=== Block {blue_objects_count} ===", 
                           (10, info_start_y), cv2.FONT_HERSHEY_SIMPLEX, 0.6, (0, 255, 255), 2)
                cv2.putText(result_frame, f"Center: ({cx}, {cy})", 
                           (10, info_start_y + 20), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                cv2.putText(result_frame, f"Rotation: {angle:.1f}°", 
                           (10, info_start_y + 35), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                cv2.putText(result_frame, f"Area: {int(area)} px²", 
                           (10, info_start_y + 50), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                cv2.putText(result_frame, f"Size: {width_rect:.0f}x{height_rect:.0f}", 
                           (10, info_start_y + 65), cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
        
        # 在图像顶部显示检测到的蓝色积木块总数
        total_info = f"Total Blue Blocks Detected: {blue_objects_count}"
        cv2.putText(result_frame, total_info, (10, result_frame.shape[0] - 20), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 0), 2)
        
        # 添加图例说明
        legend_x = result_frame.shape[1] - 250
        cv2.putText(result_frame, "Legend:", (legend_x, 30), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 255), 2)
        cv2.putText(result_frame, "Green: Contour", (legend_x, 50), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 1)
        cv2.putText(result_frame, "Blue: Min Area Rect", (legend_x, 65), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 0, 0), 1)
        cv2.putText(result_frame, "Red: Center Point", (legend_x, 80), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 0, 255), 1)
        cv2.putText(result_frame, "Yellow: Rotation", (legend_x, 95), 
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 255), 1)
        
        # 显示原始画面和处理结果
        cv2.imshow('Blue Block Detection - Enhanced', result_frame)
        cv2.imshow('Blue Mask', mask)
        
        # 检查按键输入，如果按下 'q' 键则退出
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            print("用户按下 'q' 键，退出程序")
            break
    
    # 释放摄像头资源
    cap.release()
    # 关闭所有OpenCV窗口
    cv2.destroyAllWindows()
    print("摄像头已关闭，蓝色积木块检测程序结束")

if __name__ == "__main__":
    main()
